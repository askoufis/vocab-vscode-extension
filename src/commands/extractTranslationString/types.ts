import * as vscode from "vscode";

const STRING_LITERAL_TYPES = ["regular", "jsx", "prop"] as const;
export type StringLiteralType = typeof STRING_LITERAL_TYPES[number];

export interface TranslationString {
  /** The string literal value, excluding surrounding quotes if the selection contains quotes */
  value: string;
  /** The selection containing the string literal and quotes if they exist (i.e. if it's not a JSX string literal) */
  selection: vscode.Selection;
  /** Whether or not the string literal is inside JSX*/
  type: StringLiteralType;
}
