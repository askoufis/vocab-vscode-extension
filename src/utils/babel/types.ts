import * as t from "@babel/types";

export interface ElementName {
  name: string;
  suffix: string;
}
export type ElementNameOccurrences = Record<string, number | undefined>;

export interface TransformState {
  key: string;
  message: string;
  elementNameOccurrences: ElementNameOccurrences;
  translationHookProperties: t.ObjectProperty[];
  elementNameStack: ElementName[];
}

type TransformStateOutput = Pick<TransformState, "key" | "message">;

export interface TransformResult extends TransformStateOutput {
  code: string;
}

export type OnTreeExit = (transformStateOutput: TransformStateOutput) => void;

interface PluginOptions {
  onTreeExit: OnTreeExit;
}

export interface PluginState extends TransformState {
  opts: PluginOptions;
}
