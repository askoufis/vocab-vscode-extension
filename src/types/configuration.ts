export type MaxTranslationKeyLength = number | null;
export type FormatAfterReplace = boolean;

export interface Configuration {
  maxTranslationKeyLength: MaxTranslationKeyLength;
  formatAfterReplace: FormatAfterReplace;
}
