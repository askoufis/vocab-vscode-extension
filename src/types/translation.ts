import * as vscode from "vscode";
import { TransformResult } from "../commands/extractTranslationString/utils/jsxBabelTransform";

const HIGHLIGHT_TYPES = [
  // E.g. const foo = "foo";
  // Highlight:        ___
  "stringLiteral",
  // E.g. return <div>foo</div>
  // Highlight:       ___
  "jsxStringLiteral",
  // E.g. return <Field label="foo" />
  // Highlight:                ___
  "propValueStringLiteral",
  // E.g. return <div>Foo <b>bar</b> foo</div>
  // Highlight:       __________________
  "stringLiteralAndJsx",
] as const;
export type HighlightType = typeof HIGHLIGHT_TYPES[number];

export type HighlightStringNoTransform = {
  /** The string literal value, excluding surrounding quotes if the selection contains quotes */
  value: string;
  /** The selection containing the string literal and quotes if they exist (i.e. if it's not a JSX string literal) */
  selection: vscode.Selection;
  /** What type of thing was the highlighted text */
  type: Exclude<HighlightType, "stringLiteralAndJsx">;
};

export type HighlightStringWithTransform = {
  /** The selection containing the string literal and quotes if they exist (i.e. if it's not a JSX string literal) */
  selection: vscode.Selection;
  /** What type of thing was the highlighted text */
  type: "stringLiteralAndJsx";
  /** What type of thing was the highlighted text */
  transformResult: TransformResult;
};

export type HighlightString =
  | HighlightStringNoTransform
  | HighlightStringWithTransform;

type TranslationKey = string;

interface TranslationString {
  message: string;
  description?: string;
}

export type TranslationsFile = Record<TranslationKey, TranslationString>;
