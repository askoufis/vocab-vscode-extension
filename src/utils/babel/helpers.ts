import * as t from "@babel/types";

export const createJsxElement = (
  elementName: string,
  children: t.JSXElement["children"] = []
): t.JSXElement => {
  const elementNameIdentifier = t.jsxIdentifier(elementName);
  const openingElement = t.jsxOpeningElement(elementNameIdentifier, []);
  const closingElement = t.jsxClosingElement(elementNameIdentifier);

  return t.jSXElement(openingElement, closingElement, children);
};
