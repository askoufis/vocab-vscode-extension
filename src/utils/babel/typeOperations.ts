import * as t from "@babel/types";
import { capitalise, transformWrapper } from "../string";

type JsxElementNameType = t.JSXElement["openingElement"]["name"]["type"];
interface ElementName {
  name: string;
  type: JsxElementNameType;
}

export const getJsxElementName = (jsxElement: t.JSXElement): ElementName => {
  const openingElementName = jsxElement.openingElement.name;
  if (t.isJSXIdentifier(openingElementName)) {
    return { name: openingElementName.name, type: openingElementName.type };
  }

  if (t.isJSXMemberExpression(openingElementName)) {
    const flattenedName = flattenJsxMemberExpression(openingElementName);
    const joinedName = flattenedName.map(({ name }) => name).join(".");

    return { name: joinedName, type: openingElementName.type };
  }

  // TODO: Handle namespace identifiers
  throw new Error(
    "Namespaced identifiers as JSX element names are not supported yet"
  );
};

export const isJsxVocabTransformElement = (
  jsxElement: t.JSXElement
): boolean => {
  const openingElementName = jsxElement.openingElement.name;
  if (t.isJSXIdentifier(openingElementName)) {
    return openingElementName.name === transformWrapper;
  }

  return false;
};

export const flattenMemberExpression = ({
  object,
  property,
}: t.MemberExpression): t.Identifier[] => {
  if (!t.isIdentifier(object) && !t.isMemberExpression(object)) {
    throw new Error(
      "Member expression object is not an identifier or a member expression"
    );
  }

  const flattenedObject = t.isMemberExpression(object)
    ? flattenMemberExpression(object)
    : [object];

  if (!t.isIdentifier(property)) {
    throw new Error("Member expression property is not an identifier");
  }

  return [...flattenedObject, property];
};

export const flattenJsxMemberExpression = ({
  object,
  property,
}: t.JSXMemberExpression): t.JSXIdentifier[] => {
  if (!t.isJSXIdentifier(object) && !t.isJSXMemberExpression(object)) {
    throw new Error(
      "Member expression object is not an identifier or a member expression"
    );
  }

  const flattenedObject = t.isJSXMemberExpression(object)
    ? flattenJsxMemberExpression(object)
    : [object];

  if (!t.isJSXIdentifier(property)) {
    throw new Error("Member expression property is not an identifier");
  }

  return [...flattenedObject, property];
};

export const memberExpressionToObjectProperty = (
  memberExpression: t.MemberExpression
): { objectProperty: t.ObjectProperty; keyString: string } => {
  const [first, ...rest] = flattenMemberExpression(memberExpression);
  const titleCastRest = rest.map(({ name }) => capitalise(name));
  const keyString = [first.name, ...titleCastRest].join("");

  const key = t.identifier(keyString);
  const value = memberExpression;
  const objectProperty = t.objectProperty(key, value);

  // The caller could get the key string from the object property, but it's a bit tedious
  // so it's easier to just return it alongside for now
  return { objectProperty, keyString };
};

const childrenIdentifier = t.identifier("children");

export const createElementRendererObjectProperty = (
  jsxElement: t.JSXElement
): t.ObjectProperty => {
  // Assumption: This JSXElement has no nested children, so we just replace
  // all its children with a single children identifier
  jsxElement.children = [t.jsxExpressionContainer(childrenIdentifier)];
  const { name, type } = getJsxElementName(jsxElement);

  const propertyKey =
    type === "JSXIdentifier" ? t.identifier(name) : t.stringLiteral(name);

  const arrowFunctionBody = jsxElement;
  const propertyValue = t.arrowFunctionExpression(
    [childrenIdentifier],
    arrowFunctionBody
  );

  return t.objectProperty(propertyKey, propertyValue);
};
