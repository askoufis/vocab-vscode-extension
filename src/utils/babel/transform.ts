/* eslint-disable @typescript-eslint/naming-convention */
import type { Visitor } from "@babel/core";
import { transformSync } from "@babel/core";
import * as t from "@babel/types";
import { memberExpressionToObjectProperty } from "./typeOperations";
import {
  wrapWithTransformWrapper,
  trimTrailingSemicolon,
  removeTransformWrapper,
} from "../string";
import {
  jsxElementEnterVisitor,
  jsxElementExitVisitor,
  jsxTextVisitor,
} from "./visitors";
// @ts-ignore
import babelPluginSyntaxJsx from "@babel/plugin-syntax-jsx";
// @ts-ignore
import babelPluginSyntaxTypescript from "@babel/plugin-syntax-typescript";
import type { OnTreeExit, PluginState, TransformResult } from "./types";

export const transformHighlightContainingJsx = (s: string): TransformResult => {
  // The highlighted code is likely not valid JSX by itself, so we
  // wrap it in an element babel can parse it
  const validJsxCode = wrapWithTransformWrapper(s);

  const { key, message, code } = transformJsxToVocabHook(validJsxCode);

  const trimmedCode = trimTrailingSemicolon(code);
  const unwrappedCode = removeTransformWrapper(trimmedCode);

  return { key, message, code: unwrappedCode };
};

export const transformJsxToVocabHook = (
  validJsxCode: string
): TransformResult => {
  let transformResult: TransformResult = { key: "", message: "", code: "" };

  const onTreeExit: OnTreeExit = (result) => {
    transformResult = { ...transformResult, ...result };
  };

  const code = transformSync(validJsxCode, {
    plugins: [
      babelPluginSyntaxJsx,
      [babelPluginSyntaxTypescript, { isTSX: true }],
      [vocabTransformPlugin, { onTreeExit }],
    ],
    retainLines: true,
  })?.code;

  return { ...transformResult, code: code || "" };
};

const jsxExpressionContainerVisitor: Visitor<PluginState> = {
  MemberExpression: (path, state) => {
    if (t.isJSXExpressionContainer(path.parent)) {
      const { objectProperty, keyString } = memberExpressionToObjectProperty(
        path.node
      );
      state.key = `${state.key}${keyString}`;
      state.message = `${state.message}{${keyString}}`;
      state.translationHookProperties.push(objectProperty);
    }
  },
  Identifier: ({ node, parent }, state) => {
    if (t.isJSXExpressionContainer(parent)) {
      const keyAndValue = t.identifier(node.name);
      const computed = false;
      const shorthand = true;
      const objectProperty = t.objectProperty(
        keyAndValue,
        keyAndValue,
        computed,
        shorthand
      );
      state.translationHookProperties.push(objectProperty);

      state.key = `${state.key}${node.name}`;
      state.message = `${state.message}{${node.name}}`;
    }
  },
};

const vocabTransformPlugin = (): { visitor: Visitor<PluginState> } => ({
  visitor: {
    Program: {
      enter: (_path, state) => {
        state.key = "";
        state.message = "";
        state.elementNameOccurrences = {};
        state.translationHookProperties = [];
        state.elementNameStack = [];
      },
      exit: (_path, state) => {
        state.opts.onTreeExit({
          key: state.key,
          message: state.message,
        });
      },
    },
    JSXElement: {
      enter: jsxElementEnterVisitor,
      exit: jsxElementExitVisitor,
    },
    JSXText: jsxTextVisitor,
    JSXExpressionContainer: (path, state) => {
      if (t.isJSXElement(path.parent)) {
        path.traverse(jsxExpressionContainerVisitor, state);
      }
    },
  },
});
