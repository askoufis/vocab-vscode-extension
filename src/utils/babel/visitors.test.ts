import * as t from "@babel/types";
import { createJsxElement } from "./helpers";
import { TransformState } from "./transform";
import { createElementRendererObjectProperty } from "./typeOperations";
import {
  jsxElementEnterVisitor,
  jsxElementExitVisitor,
  jsxTextVisitor,
} from "./visitors";

const initialTransformState: TransformState = {
  key: "Existing text ",
  message: "Existing text ",
  elementNameOccurrences: {},
  translationHookProperties: [],
  elementNameStack: [],
};

describe("JSXText visitor", () => {
  it("should append the text to both the key and the message", () => {
    const jsxTextNode = t.jSXText("test");
    const state = {
      ...initialTransformState,
    };

    jsxTextVisitor({ node: jsxTextNode }, state);

    expect(state).toEqual({
      key: "Existing text test",
      message: "Existing text test",
      elementNameOccurrences: {},
      translationHookProperties: [],
      elementNameStack: [],
    });
  });
});

describe("JSXElement", () => {
  describe("enter visitor", () => {
    describe("when the element is the VocabTransform element", () => {
      it("should not alter state if we the element is the VocabTransform element", () => {
        const jsxElement = createJsxElement("VocabTransform");
        const state = {
          ...initialTransformState,
        };

        jsxElementEnterVisitor({ node: jsxElement }, state);

        expect(state).toEqual(initialTransformState);
      });
    });

    describe("when the element is not the VocabTransform element", () => {
      it("push the element name onto the element name stack and append and an element tag to the message", () => {
        const jsxElement = createJsxElement("button");
        const state = {
          ...initialTransformState,
        };

        jsxElementEnterVisitor({ node: jsxElement }, state);

        expect(state).toEqual({
          key: "Existing text ",
          message: "Existing text <button>",
          elementNameOccurrences: { button: 1 },
          translationHookProperties: [],
          elementNameStack: [{ name: "button", suffix: "" }],
        });
      });
    });

    describe("when an element is visited that has the same name as a previously visited element", () => {
      it("push the element name onto the element name stack and append and an element tag to the message", () => {
        const jsxElement = createJsxElement("button");
        const state = {
          key: "Existing text <button>foo</button>",
          message: "Existing text <button>foo</button> ",
          elementNameOccurrences: { button: 1 },
          translationHookProperties: [],
          elementNameStack: [{ name: "button", suffix: "" }],
        };

        jsxElementEnterVisitor({ node: jsxElement }, state);

        expect(state).toEqual({
          key: "Existing text <button>foo</button>",
          message: "Existing text <button>foo</button> <button1>",
          elementNameOccurrences: { button: 2 },
          translationHookProperties: [],
          elementNameStack: [
            { name: "button", suffix: "" },
            { name: "button", suffix: "1" },
          ],
        });
      });
    });
  });

  describe("exit visitor", () => {
    describe("when the element is the VocabTransform element", () => {
      it("should replace its children with the translation hook call", () => {
        const jsxElement = createJsxElement("VocabTransform");
        const state: TransformState = {
          key: "Click here",
          message: "Click <button>here</button",
          elementNameOccurrences: {},
          translationHookProperties: [
            createElementRendererObjectProperty(createJsxElement("button")),
          ],
          elementNameStack: [],
        };

        jsxElementExitVisitor({ node: jsxElement }, state);

        expect(jsxElement.children).toMatchSnapshot();
      });
    });

    describe("when the element is not the VocabTransform element", () => {
      it("should append its closing tag to the message and add its render function to the translation hook properties", () => {
        const elementName = "button";
        const jsxElement = createJsxElement(elementName, [
          t.jSXText("click me"),
        ]);
        const state: TransformState = {
          key: "Existing text click me",
          message: "Existing text <button>click me",
          elementNameOccurrences: {},
          translationHookProperties: [],
          elementNameStack: [{ name: elementName, suffix: "" }],
        };

        jsxElementExitVisitor({ node: jsxElement }, state);

        expect(state).toMatchSnapshot();
      });
    });
  });
});
