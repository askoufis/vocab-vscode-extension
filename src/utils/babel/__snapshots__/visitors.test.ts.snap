// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JSXElement exit visitor when the element is not the VocabTransform element should append its closing tag to the message and add its render function to the translation hook properties 1`] = `
Object {
  "elementNameOccurrences": Object {},
  "elementNameStack": Array [],
  "key": "Existing text click me",
  "message": "Existing text <button>click me</button>",
  "translationHookProperties": Array [
    Object {
      "computed": false,
      "decorators": null,
      "key": Object {
        "name": "button",
        "type": "Identifier",
      },
      "shorthand": false,
      "type": "ObjectProperty",
      "value": Object {
        "async": false,
        "body": Object {
          "children": Array [
            Object {
              "expression": Object {
                "name": "children",
                "type": "Identifier",
              },
              "type": "JSXExpressionContainer",
            },
          ],
          "closingElement": Object {
            "name": Object {
              "name": "button",
              "type": "JSXIdentifier",
            },
            "type": "JSXClosingElement",
          },
          "openingElement": Object {
            "attributes": Array [],
            "name": Object {
              "name": "button",
              "type": "JSXIdentifier",
            },
            "selfClosing": false,
            "type": "JSXOpeningElement",
          },
          "selfClosing": null,
          "type": "JSXElement",
        },
        "params": Array [
          Object {
            "name": "children",
            "type": "Identifier",
          },
        ],
        "type": "ArrowFunctionExpression",
      },
    },
  ],
}
`;

exports[`JSXElement exit visitor when the element is the VocabTransform element should replace its children with the translation hook call 1`] = `
Array [
  Object {
    "expression": Object {
      "arguments": Array [
        Object {
          "type": "StringLiteral",
          "value": "Click here",
        },
        Object {
          "properties": Array [
            Object {
              "computed": false,
              "decorators": null,
              "key": Object {
                "name": "button",
                "type": "Identifier",
              },
              "shorthand": false,
              "type": "ObjectProperty",
              "value": Object {
                "async": false,
                "body": Object {
                  "children": Array [
                    Object {
                      "expression": Object {
                        "name": "children",
                        "type": "Identifier",
                      },
                      "type": "JSXExpressionContainer",
                    },
                  ],
                  "closingElement": Object {
                    "name": Object {
                      "name": "button",
                      "type": "JSXIdentifier",
                    },
                    "type": "JSXClosingElement",
                  },
                  "openingElement": Object {
                    "attributes": Array [],
                    "name": Object {
                      "name": "button",
                      "type": "JSXIdentifier",
                    },
                    "selfClosing": false,
                    "type": "JSXOpeningElement",
                  },
                  "selfClosing": null,
                  "type": "JSXElement",
                },
                "params": Array [
                  Object {
                    "name": "children",
                    "type": "Identifier",
                  },
                ],
                "type": "ArrowFunctionExpression",
              },
            },
          ],
          "type": "ObjectExpression",
        },
      ],
      "callee": Object {
        "name": "t",
        "type": "Identifier",
      },
      "type": "CallExpression",
    },
    "type": "JSXExpressionContainer",
  },
]
`;
